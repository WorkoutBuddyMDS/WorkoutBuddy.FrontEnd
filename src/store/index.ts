import { configureStore } from '@reduxjs/toolkit';
import { accountReducer } from './reducers/account';

const store = configureStore({
  reducer: {
    account: accountReducer,
  },
});

export type RootState = ReturnType<typeof store.getState>;

export default store;
