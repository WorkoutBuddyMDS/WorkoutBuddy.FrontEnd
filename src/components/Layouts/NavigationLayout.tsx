import * as React from 'react';
import AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import IconButton from '@mui/material/IconButton';
import Typography from '@mui/material/Typography';
import Menu from '@mui/material/Menu';
import MenuIcon from '@mui/icons-material/Menu';
import Container from '@mui/material/Container';
import Avatar from '@mui/material/Avatar';
import Button from '@mui/material/Button';
import Tooltip from '@mui/material/Tooltip';
import MenuItem from '@mui/material/MenuItem';
import AdbIcon from '@mui/icons-material/Adb';
import { StyledBasicButton, StyledLink } from '../../styles/styled-components';
import { styled } from '@mui/system';
import { useRouter } from 'next/router';
import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { FitnessCenter } from '@mui/icons-material';
import Link from 'next/link';
import { accountActions } from '../../store/reducers/account';
import { RootState } from '../../store';
import LanguageSwitch from '../../components/LanguageSwitch/LanguageSwitch';
import useText from '../../services/site-properties/parsing';

const pages = [
  <StyledLink href="/exercises">Exercises</StyledLink>,
  <StyledLink href="/splits">Splits</StyledLink>,
  <StyledLink href="/caloriesCalculator">Calories Calculator</StyledLink>,
];

const StyledButtonBox = styled(Box)`
  margin-left: 10px;
`;

const NavigationLayout = React.forwardRef((props: any) => {
  const router = useRouter();
  const locale = useSelector((state: RootState) => state.language.language);
  const dispatcher = useDispatch();
  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const accountState = useSelector((state: RootState) => state.account);
  const [username, setUsername] = useState(accountState.username);
  const [isAdmin, setisAdmin] = useState<boolean | undefined>(false);
  const [anchorElNav, setAnchorElNav] = React.useState<null | HTMLElement>(
    null
  );
  const [anchorElUser, setAnchorElUser] = React.useState<null | HTMLElement>(
    null
  );
  const text = {
    exercisesTab: useText('navigation.layout.tabs.exercises.text', locale),
    splitsTab: useText('navigation.layout.tabs.splits.text', locale),
    signInBtn: useText('navigation.layout.buttons.signin.text', locale),
    registerBtn: useText('navigation.layout.buttons.register.text', locale),
    profileText: useText('navigation-layout.profile.button', locale),
    adminText: useText('navigation-layout.admin.button', locale),
    logoutText: useText('navigation-layout.logout.button', locale),
    calorieCalculator: useText('pages.calorie-calculator.title.header', locale),
  };

  const pages = [
    <StyledLink href="/exercises" key={1}>
      {text.exercisesTab}
    </StyledLink>,
    <StyledLink href="/splits" key={2}>
      {text.splitsTab}
    </StyledLink>,
    <StyledLink href="/caloriesCalculator" key={3}>
      {text.calorieCalculator}
    </StyledLink>,
  ];

  useEffect(() => {
    const token = sessionStorage.getItem('token');
    if (token) {
      setIsLoggedIn(true);
      setUsername(sessionStorage.getItem('username') ?? 'User');
      setisAdmin(sessionStorage.getItem('roles')?.includes('Admin'));
    } else {
      setIsLoggedIn(false);
    }
  }, [accountState]);

  const handleOpenNavMenu = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorElNav(event.currentTarget);
  };
  const handleOpenUserMenu = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorElUser(event.currentTarget);
  };

  const handleCloseNavMenu = () => {
    setAnchorElNav(null);
  };

  const handleCloseUserMenu = () => {
    setAnchorElUser(null);
  };

  const settings = [
    <Link href={`/user/edit`} key={1}>
      {text.profileText}
    </Link>,
    <Button
      key={2}
      onClick={() => {
        dispatcher(accountActions.signOut());
        router.reload();
      }}
    >
      {text.logoutText}
    </Button>,
  ];

  if (isAdmin) {
    settings.unshift(
      <Link href={`/admin/pending-exercises`}>{text.adminText}</Link>
    );
  }

  return (
    <div className="navigation-layout">
      <AppBar position="static" data-testid="navigation-layout">
        <Container maxWidth={false}>
          <Toolbar disableGutters>
            <FitnessCenter
              sx={{ display: { xs: 'none', md: 'flex' }, mr: 1 }}
            />
            <Typography
              variant="h5"
              noWrap
              component="a"
              href="/"
              sx={{
                mr: 2,
                display: { xs: 'none', md: 'flex' },
                fontFamily: 'monospace',
                fontWeight: 700,
                letterSpacing: '.15rem',
                color: 'inherit',
                textDecoration: 'none',
              }}
            >
              Workout Buddy
            </Typography>

            <Box sx={{ flexGrow: 1, display: { xs: 'flex', md: 'none' } }}>
              <IconButton
                size="large"
                aria-label="account of current user"
                aria-controls="menu-appbar"
                aria-haspopup="true"
                onClick={handleOpenNavMenu}
                color="inherit"
              >
                <MenuIcon />
              </IconButton>
              <Menu
                id="menu-appbar"
                anchorEl={anchorElNav}
                anchorOrigin={{
                  vertical: 'bottom',
                  horizontal: 'left',
                }}
                keepMounted
                transformOrigin={{
                  vertical: 'top',
                  horizontal: 'left',
                }}
                open={Boolean(anchorElNav)}
                onClose={handleCloseNavMenu}
                sx={{
                  display: { xs: 'block', md: 'none' },
                }}
              >
                {typeof window !== 'undefined' &&
                  sessionStorage.getItem('token') &&
                  pages.map((page, index) => (
                    <Typography textAlign="center" key={index}>
                      {page}
                    </Typography>
                  ))}
              </Menu>
            </Box>
            <AdbIcon sx={{ display: { xs: 'flex', md: 'none' }, mr: 1 }} />
            <Typography
              variant="h5"
              noWrap
              component="a"
              href=""
              sx={{
                mr: 2,
                display: { xs: 'flex', md: 'none' },
                flexGrow: 1,
                fontFamily: 'monospace',
                fontWeight: 700,
                letterSpacing: '.3rem',
                color: 'inherit',
                textDecoration: 'none',
              }}
            >
              Workout Buddy
            </Typography>
            <Box sx={{ flexGrow: 1, display: { xs: 'none', md: 'flex' } }}>
              {typeof window !== 'undefined' &&
                sessionStorage.getItem('token') &&
                pages.map((page, index) => (
                  <Typography
                    sx={{ my: 2, color: 'white', display: 'block' }}
                    key={index}
                  >
                    {page}
                  </Typography>
                ))}
            </Box>

            <LanguageSwitch />

            {isLoggedIn ? (
              <Box sx={{ flexGrow: 0 }}>
                <Tooltip title="Open settings">
                  <IconButton onClick={handleOpenUserMenu} sx={{ p: 0 }}>
                    <Avatar>
                      {username.replace(/[a-z]/g, '') ||
                        username.toUpperCase()[0]}
                    </Avatar>
                  </IconButton>
                </Tooltip>
                <Menu
                  sx={{ mt: '45px' }}
                  id="menu-appbar"
                  anchorEl={anchorElUser}
                  anchorOrigin={{
                    vertical: 'top',
                    horizontal: 'right',
                  }}
                  keepMounted
                  transformOrigin={{
                    vertical: 'top',
                    horizontal: 'right',
                  }}
                  open={Boolean(anchorElUser)}
                  onClose={handleCloseUserMenu}
                >
                  {settings.map((setting, index) => (
                    <MenuItem onClick={handleCloseUserMenu} key={index}>
                      <Typography textAlign="center">{setting}</Typography>
                    </MenuItem>
                  ))}
                </Menu>
              </Box>
            ) : (
              <StyledButtonBox>
                <StyledBasicButton onClick={() => router.push('/login')}>
                  {text.signInBtn}
                </StyledBasicButton>
                <StyledBasicButton onClick={() => router.push('/register')}>
                  {text.registerBtn}
                </StyledBasicButton>
              </StyledButtonBox>
            )}
          </Toolbar>
        </Container>
      </AppBar>
      {props.children}
    </div>
  );
});

NavigationLayout.displayName = 'NavigationLayout';
export default NavigationLayout;
