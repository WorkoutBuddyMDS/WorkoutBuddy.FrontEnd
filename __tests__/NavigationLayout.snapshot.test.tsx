import * as React from 'react';
import NavigationLayout from '../src/components/Layouts/NavigationLayout';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import store from '../src/store';
import '@testing-library/jest-dom';

const useRouter = jest.spyOn(require('next/router'), 'useRouter');

useRouter.mockImplementation(() => ({
  route: '/',
  pathname: '',
  query: '',
  asPath: '',
  push: jest.fn(),
  replace: jest.fn(),
  reload: jest.fn(),
  back: jest.fn(),
  prefetch: jest.fn(),
  beforePopState: jest.fn(),
  isFallback: false,
}));

describe('Tests', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Provider store={store}>
        <ForwardRefWrapper />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  it('expects correct text to be displayed', async () => {
    const { getByText } = render(
      <Provider store={store}>
        <ForwardRefWrapper />
      </Provider>
    );

    expect(getByText('test')).toBeInTheDocument();
  });
});

const ForwardRefWrapper = React.forwardRef((props, ref) => (
  <NavigationLayout ref={ref} {...props}>
    <div className="default-child-text">test</div>
  </NavigationLayout>
));
