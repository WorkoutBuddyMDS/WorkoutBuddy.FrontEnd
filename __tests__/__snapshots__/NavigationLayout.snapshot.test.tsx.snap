// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests renders correctly 1`] = `
<div>
  <div
    class="navigation-layout"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-hip9hq-MuiPaper-root-MuiAppBar-root"
      data-testid="navigation-layout"
    >
      <div
        class="MuiContainer-root css-ur2jdm-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-regular css-r6ewbb-MuiToolbar-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1uzrjc8-MuiSvgIcon-root"
            data-testid="FitnessCenterIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20.57 14.86 22 13.43 20.57 12 17 15.57 8.43 7 12 3.43 10.57 2 9.14 3.43 7.71 2 5.57 4.14 4.14 2.71 2.71 4.14l1.43 1.43L2 7.71l1.43 1.43L2 10.57 3.43 12 7 8.43 15.57 17 12 20.57 13.43 22l1.43-1.43L16.29 22l2.14-2.14 1.43 1.43 1.43-1.43-1.43-1.43L22 16.29z"
            />
          </svg>
          <a
            class="MuiTypography-root MuiTypography-h5 MuiTypography-noWrap css-ot2t7s-MuiTypography-root"
            href="/"
          >
            Workout Buddy
          </a>
          <div
            class="MuiBox-root css-1lvtzne"
          >
            <button
              aria-controls="menu-appbar"
              aria-haspopup="true"
              aria-label="account of current user"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeLarge css-1pz7awu-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="MenuIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-z5gn4u-MuiSvgIcon-root"
            data-testid="AdbIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M5 16c0 3.87 3.13 7 7 7s7-3.13 7-7v-4H5v4zM16.12 4.37l2.1-2.1-.82-.83-2.3 2.31C14.16 3.28 13.12 3 12 3s-2.16.28-3.09.75L6.6 1.44l-.82.83 2.1 2.1C6.14 5.64 5 7.68 5 10v1h14v-1c0-2.32-1.14-4.36-2.88-5.63zM9 9c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1zm6 0c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1z"
            />
          </svg>
          <a
            class="MuiTypography-root MuiTypography-h5 MuiTypography-noWrap css-1lfs2tq-MuiTypography-root"
            href=""
          >
            Workout Buddy
          </a>
          <div
            class="MuiBox-root css-1t6c9ts"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-4oy5zj-MuiTypography-root"
            >
              <a
                class="css-14yxq3c"
                href="/exercises"
              >
                Exerciții
              </a>
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-4oy5zj-MuiTypography-root"
            >
              <a
                class="css-14yxq3c"
                href="/splits"
              >
                Programe
              </a>
            </p>
          </div>
          <span>
            <button
              class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-default MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-default css-j5omkc-MuiButtonBase-root-MuiFab-root"
              tabindex="0"
              type="button"
            >
              <svg
                id="flag-icons-gb"
                viewBox="0 0 640 480"
                width="40"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h640v480H0z"
                  fill="#012169"
                />
                <path
                  d="m75 0 244 181L562 0h78v62L400 241l240 178v61h-80L320 301 81 480H0v-60l239-178L0 64V0h75z"
                  fill="#FFF"
                />
                <path
                  d="m424 281 216 159v40L369 281h55zm-184 20 6 35L54 480H0l240-179zM640 0v3L391 191l2-44L590 0h50zM0 0l239 176h-60L0 42V0z"
                  fill="#C8102E"
                />
                <path
                  d="M241 0v480h160V0H241zM0 160v160h640V160H0z"
                  fill="#FFF"
                />
                <path
                  d="M0 193v96h640v-96H0zM273 0v480h96V0h-96z"
                  fill="#C8102E"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </span>
          <div
            class="MuiBox-root css-1ruocjc"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-pklmf0-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Conectare
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-pklmf0-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Înregistrare
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </header>
    <div
      class="default-child-text"
    >
      test
    </div>
  </div>
</div>
`;
